File: 940-distinct-subsequences-ii/Solution.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        MOD = 10**9 + 7

        last = [0] * 26

        dp = 1  # Empty subsequence

        for ch in s:
            index = ord(ch) - ord('a')

            new_dp = (2 * dp - last[index]) % MOD

            last[index] = dp
            dp = new_dp

        # Remove the empty subsequence
        return (dp - 1) % MOD

File: 940-distinct-subsequences-ii/Solution.java
class Solution {
    public int distinctSubseqII(String s) {
        final int MOD = 1_000_000_007;

        long[] last = new long[26];

        long dp = 1; // Empty subsequence

        for (char ch : s.toCharArray()) {
            int index = ch - 'a';

            long newDp = (2 * dp - last[index] + MOD) % MOD;

            last[index] = dp;
            dp = newDp;
        }

        // Remove the empty subsequence
        return (int) ((dp - 1 + MOD) % MOD);
    }
}
